### PR TITLE
write_json() should write both Array[_] and Objects [BW-402]

### PR DESCRIPTION
**Pull request by salonishah11**
Closes https://broadworkbench.atlassian.net/browse/BW-402

### CODE REVIEW

File: cwl/src/test/scala/cwl/internal/EnhancedRhinoSandboxSpec.scala
Comment (salonishah11) [inline cwl/src/test/scala/cwl/internal/EnhancedRhinoSandboxSpec.scala:69]: I updated this as compiler was complaining that `.newInstance` is deprecated

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/action/ActionUtils.scala
Comment (salonishah11) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/action/ActionUtils.scala:13]: Similarly here the compiler was complaining that `new Integer(n)` is deprecated and `Integer.valueOf(n)` should be used instead

File: centaur/src/main/resources/standardTestCases/write_json/write_json.wdl
Comment (aednichols) [inline centaur/src/main/resources/standardTestCases/write_json/write_json.wdl:5]:
```suggestion
        Array[Int] indices = [0, 1, 2]
```

File: centaur/src/main/resources/standardTestCases/write_json/write_json.wdl
Comment (aednichols) [inline centaur/src/main/resources/standardTestCases/write_json/write_json.wdl:30]: I endorse this

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/types/EngineFunctionEvaluators.scala
Comment (aednichols) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/types/EngineFunctionEvaluators.scala:143]: Intentional bitwise OR? Should work fine but meaning unclear.
```suggestion
        case v if WomArrayType(WomAnyType).isCoerceableFrom(v) || WomObjectType.isCoerceableFrom(v) => v.validNel
```

**Review by aednichols — APPROVED** [review]
Nicely done! I am surprised and pleased the fix for this is so concise 👍 

It would be nice to have an identical test for `version development` because that version will be under active development as part of WDL 2.0 support

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/types/EngineFunctionEvaluators.scala
Comment (salonishah11) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/types/EngineFunctionEvaluators.scala]: Sorry this was not intentional. I have updated it to `||`

**Review by cjllanwarne — CHANGES_REQUESTED** [review]
I know the ticket only asked for `Array`s, but the spec actually says we should be able to handle "any type" (https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#file-write_jsonmixed).

I guess I'll leave that up to you whether to add in the other cases or leave it at objects and arrays for now.

Only required change is to the result type of the TypeEvaluator.

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/types/EngineFunctionEvaluators.scala
Comment (salonishah11) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/types/EngineFunctionEvaluators.scala:143]: Good point. I missed that we mapped it to `WomSingleFileType` previously. I will update it.

File: centaur/src/main/resources/standardTestCases/write_json/write_json.wdl
Comment (salonishah11) [reply centaur/src/main/resources/standardTestCases/write_json/write_json.wdl:16]: Great idea!

**Comment by salonishah11** [comment]
@cjllanwarne I was not aware that `write_json()` was expected to support all types listed [here](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#mixed-read_jsonstringfile). If that's what is in the spec I think we should support it. I will make the necessary changes and request re-reviews.

@aednichols FYI ^

File: CHANGELOG.md
Comment (aednichols) [inline CHANGELOG.md:8]:
```suggestion
* Fixed a bug that prevented `write_json()` from working with arrays of objects. The function now works as expected for `Boolean`, `String`, `Integer`, `Object` and `Array[_]` 
(including array of objects) type inputs. More information on WDL Type to JSON Type conversion can be found [here](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#mixed-read_jsonstringfile).
```
Pure ToL

**Review by aednichols — APPROVED** [review]
Even better with the additional type support! 👍

File: CHANGELOG.md
Comment (salonishah11) [reply CHANGELOG.md:8]: Oooh this is so much better. Thank you! :)

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/types/EngineFunctionEvaluators.scala
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/types/EngineFunctionEvaluators.scala:149]:
```suggestion
          s"'Array[_]' but got '${v.friendlyName}'").invalidNel
```

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/types/EngineFunctionEvaluators.scala
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/types/EngineFunctionEvaluators.scala:144]: We probably also want to allow `WomFloat` (and possibly Pair and Map)?

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:344]: These two coercions feel unnecessary?

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:346]: Same comment about probably wanting `Float` and possibly wanting `Map` and `Pair` here.

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala
Comment (cjllanwarne) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:342]: You might be able to cut down the number of cases here by saying `case v: WomPrimitive => convertToSingleFile(v)`

File: CHANGELOG.md
Comment (salonishah11) [reply CHANGELOG.md:7]: I can add support for `Float` and `Pair[_, _]`. I thought the `WomObjectType` covers `Map[_, _]` as well?

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala
Comment (salonishah11) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:344]: I will remove them

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala
Comment (salonishah11) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:342]: I am afraid of using `case v: WomPrimitive => convertToSingleFile(v)` since this means types like `WomEnumerationValue`, `WomGlobFile`, `WomSingleFile` will also be valid for `write_json()`

File: CHANGELOG.md
Comment (cjllanwarne) [inline CHANGELOG.md:3]: Just NB if we don't merge this in time for the release today, you might need to move the comment

File: centaur/src/main/resources/standardTestCases/write_json/write_json_version_develop.wdl
Comment (cjllanwarne) [inline centaur/src/main/resources/standardTestCases/write_json/write_json_version_develop.wdl:15]: TOL (not required, just curious): Is there a reason you have to do this as a call rather than 

```wdl
scatter (i in indices) {
  Random randoms = object {index: i, name: "mr_bean"}
}

[...]

output {
        String object_out = read_string(write_json(randoms[0]))
        String array_object_out = read_string(write_json(randoms))
}
```

**Review by cjllanwarne — APPROVED** [review]
LGTM. One minor question about the test case but not a blocker to merging and fixing later (if you do want to change it, it'd be nice to get this merged before the release and then address that in a follow-up PR)

File: centaur/src/main/resources/standardTestCases/write_json/write_json_version_develop.wdl
Comment (salonishah11) [reply centaur/src/main/resources/standardTestCases/write_json/write_json_version_develop.wdl:15]: To be honest, initially I was trying to have `task`s in the WDL. But this is better. No tasks, no Papi calls, no VMs. Once this PR is merged I will create another PR with these changes. Thank you!

**Pull request merged**
